README.md
# ZABCOMBAIN #
## About
This program helps you to generate reports, and adds the results of the ping command.
For example, we get such a results:

```
name     :Zabbix server
hostname :Zabbix server

ZABBIX PING RESULT

PING 127.0.0.1 (127.0.0.1) 56(84) bytes of data.
64 bytes from 127.0.0.1: icmp_seq=1 ttl=64 time=0.000 ms
64 bytes from 127.0.0.1: icmp_seq=2 ttl=64 time=0.774 ms
64 bytes from 127.0.0.1: icmp_seq=3 ttl=64 time=0.783 ms

--- 127.0.0.1 ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2002ms
rtt min/avg/max/mdev = 0.000/0.519/0.783/0.367 ms

INTERFACES PING RESULT

PING 127.0.0.1 (127.0.0.1) 56(84) bytes of data.
64 bytes from 127.0.0.1: icmp_seq=1 ttl=64 time=0.052 ms
64 bytes from 127.0.0.1: icmp_seq=2 ttl=64 time=0.078 ms
64 bytes from 127.0.0.1: icmp_seq=3 ttl=64 time=0.040 ms

--- 127.0.0.1 ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2025ms
rtt min/avg/max/mdev = 0.040/0.056/0.078/0.018 ms

```
pypi version 2.0.1

## Dependences
* Python >=3
* Tested against Zabbix 1.8 through 3.4
* Python3-dev packages for linux

## Dependent packages in pip
* pyzabbix - automatically installs
* PyQt5 - can be manually set

For Linux, Windows and MacOS
> To install PyQt5, type the command:
```
pip install PyQt5
or
pip install -r requirements.txt
```

## Documentation ##
### Getting Started

* Download the `zip` file and unzip it. Go to the file folder `setup.py`.
* Or use `git` to clone the ** zabcombain** project. Go to the folder with the file `setup.py`.

In the console
```
$ python3 setup.py install
```
Run util:
```
$ zabcombain
```
#### Main page:
> * To connect to the server, you need to ***'settings'*** specify the server address.
> * Then enter the username and password to work.
> * After a successful connection, is displayed in the console ***connect***

#### Oreder page:
> * To generate the event data, copy data from ***dashboard - action***.

```
Z. server Disk I/O is overloaded on server Z. No 3d 23h 13m
```
> * Click ***'paste'*** to paste the data.
> * Click ***'proccess'*** that the data started to be processed.
> * Click ***'copy'*** to copy the processed data.
> * The Tick ***'Contin.'*** , allows to clear the fields of the input/output.

#### The Ping page

> * Click ***'update'*** .
> * Select ***'host group'*** .
> * Select ***'hosts'*** .
> * Click ***'ping'*** .
> * The program ping all interfaces on the host, the results are displayed in the console.


## License ##
GPL v.3
http://www.gnu.org/licenses/gpl-3.0.html


changes.txt

-------------------------------------------------------------------------------
                             zabcombain Change Log
-------------------------------------------------------------------------------

2.0.0: (released 2018-05-20)
----------------------------

INCOMPATIBLE CHANGES SINCE 2.0.0:

- Переход на фреймворк PyQt5
- Изменение название методов в zabmodule.py
- Добавление класса TaskRuner для работы с потоками


1.0.3: (released 2018-02-24)
----------------------------

INCOMPATIBLE CHANGES SINCE 1.0.3:

- Изменена обработка списка интерфейсов в zabmodule.PingRuner(). Добавлены
функции zabmodule.changeInterfaceList(), zabmodule.checkingList(). Функция
zabmodule.changeInterfaceList() изменяет список интерфейсов так что бы в
списке не было интерфейсов с одинаковым IP, если тип интерфейса где 'main'
равно единице имеет наименьшее значение, то он выносится вперед списка.


tests/zab.py
from zabmodule import *

hosts = []


for i in range(10):
    host = initHost()
    ifs = [{'ip':'127.0.0.1'},{'ip':'localhost'},{'ip':'192.168.56.101'}]
    host.update([("name",'twigl_'+str(i))])
    host.update([("ip",'localhost')])
    host.update([('interfaces',ifs)])
    hosts.append(host)
# pr = PingRuner_()
# pr.run(hosts)

pingresult = None

def do_end():
    print('task -- end')

def iface_work(item):
    res = pingFromOS(item['ip']).decode()
    item.update([('res',res)])

def host_work(item):
    print(item.get("name"))
    ifacesTasks = TaskRuner()
    ifacesTasks.items = item['interfaces']
    ifacesTasks.do_work = iface_work
    ifacesTasks.run()


runHosts = TaskRuner()
runHosts.items = hosts
runHosts.do_work = host_work
runHosts.do_end =do_end
runHosts.run()

for host in hosts:
    print(host['name'])
    for item in host['interfaces']:
        print(item['res'])


tests/pingqt.py
from PyQt5.QtWidgets import *
from PyQt5.QtCore import *
from subprocess import Popen,PIPE
import sys


lock = QReadWriteLock()

class Worker(QThread):
    finished = pyqtSignal(str)
    def __init__(self,parent = None ):
        super(Worker,self).__init__(parent)

    def setIP(self,ip):
        self.ip = ip

    def pingFromOS(self,ip):
        if ip is None:
            return b"ip is None"
        if ip=='':
            return b'ip is empty'
        pcArg=['ping','-c','9',ip]
        proc = Popen(pcArg,stdout=PIPE)
        out = proc.stdout.readlines()
        res = b""
        for row in out:
            res+=row
        proc.stdout.close()
        return res.decode()


    def run(self):
        res  = self.pingFromOS(self.ip)
        self.finished.emit(res)



class MainWindow(QMainWindow):
    def __init__(self,*args,**kwargs):
        super(MainWindow,self).__init__(*args,**kwargs)
        self.initUI()

    def initUI(self):
        self.width = 600
        self.height = 600
        self.resize(self.width,self.height)
        self.center()
        self.ipField = QLineEdit(self)
        self.ipField.setText('127.1.1.0')
        self.ipField.setGeometry(10,10,100,25)
        self.button = QPushButton("ping",self)
        self.button.setGeometry(120,10,100,25)
        self.button.clicked.connect(self.onPressPing)
        self.console = QPlainTextEdit(self)
        self.console.setGeometry(5,50,self.width-10,self.height-60)
        self.setWindowTitle('Ping from os')
        self.show()

    def onPressPing(self):
        self.w = Worker()
        self.w.setIP(self.ipField.text())
        self.w.start()
        self.w.finished.connect(self.OnPing)


    def OnPing(self,res):
        self.console.appendPlainText(res)

    def center(self):
        qr = self.frameGeometry()
        cp = QDesktopWidget().availableGeometry().center()
        qr.moveCenter(cp)
        self.move(qr.topLeft())



if __name__ == '__main__':
    # Next, create an application object.
    QApplication.setDesktopSettingsAware(False)
    app = QApplication(sys.argv)
    # Show it.
    frame = MainWindow()
    # Start the event loop.
    sys.exit(app.exec_())


tests/testzabmodule.py
import unittest
from unittest.mock import MagicMock
from zabcombain import zabmodule

class TestMethod(unittest.TestCase):
    """docstring for TestMethod."""
    # def __init__(self,args):
    #     super(TestMethod, self).__init__(args)

    def setupApi(self):
        self.api = MagicMock()
        self.api.script = MagicMock()
        self.scriptValue = 'result value'
        self.api.script.execute = MagicMock(return_value={'value':self.scriptValue})


    def setUp(self):
        self.setupApi()
        ###############################
        self.host = zabmodule.initHost()
        ##################################################
        self.lhost = 'localhost'
        self.iphost = '127.0.0.1'
        ##############   add TESTLIST1 ###################
        self.TESTLIST1 = ({'hostid': '3425','main': '1','type': '1','ip': '127.0.0.1'},
        {'hostid': '342','main': '1','type': '2','ip': '127.0.0.1'},
        {'hostid': '425','main': '1','type': '3','ip': '127.0.0.1'},
        {'hostid': '442','main': '1','type': '4','ip': '127.0.0.1'},
        {'hostid': '465','main': '0','type': '1','ip': '127.0.0.1'},
        {'hostid': '405','main': '0','type': '2','ip': '127.0.0.1'},
        {'hostid': '135','main': '0','type': '3','ip': '127.0.0.1'},
        {'hostid': '245','main': '0','type': '4','ip': '127.0.0.1'}
        )
        ##############   add TESTLIST2 ###################
        self.TESTLIST2 = ({'hostid': '3425','main': '1','type': '1','ip': '127.0.0.1'},
        {'hostid': '342','main': '1','type': '2','ip': '127.0.0.2'},
        {'hostid': '425','main': '1','type': '3','ip': '127.0.0.3'},
        {'hostid': '442','main': '1','type': '4','ip': '127.0.0.4'},
        {'hostid': '465','main': '0','type': '1','ip': '127.0.0.1'},
        {'hostid': '405','main': '0','type': '2','ip': '127.0.0.2'},
        {'hostid': '135','main': '0','type': '3','ip': '127.0.0.3'},
        {'hostid': '245','main': '0','type': '4','ip': '127.0.0.4'}
        )


    def testCheckingList(self):
        self.assertEqual(zabmodule.checkingList(self.TESTLIST1)[0], {'hostid': '3425','main': '1','type': '1','ip': '127.0.0.1'})
        res = [{'hostid': '3425','main': '1','type': '1','ip': '127.0.0.1'},
        {'hostid': '342','main': '1','type': '2','ip': '127.0.0.2'},
        {'hostid': '425','main': '1','type': '3','ip': '127.0.0.3'},
        {'hostid': '442','main': '1','type': '4','ip': '127.0.0.4'}]
        self.assertEqual(zabmodule.checkingList(self.TESTLIST2), res)

    def testChangeInterfaceList(self):
        self.assertEqual(zabmodule.changeInterfaceList(self.TESTLIST1), [{'hostid': '3425','main': '1','type': '1','ip': '127.0.0.1'}])

    @unittest.skip("skip ping from OS")
    def testPingFromOS(self):
        self.assertEqual(zabmodule.pingFromOS(None),b"ip is None")
        self.assertEqual(zabmodule.pingFromOS(""),b"ip is empty")
        l = len(zabmodule.pingFromOS(self.iphost).decode().splitlines())
        self.assertEqual(l,8)

    def testFromRemoteServerScript(self):
        res = zabmodule.runRemoteServerScript(None,self.host)
        self.assertEqual(res,'api is None')
        res = zabmodule.runRemoteServerScript(self.api,None)
        self.assertEqual(res,'host is None')
        res = zabmodule.runRemoteServerScript(self.api,self.host)
        self.assertEqual(self.host.get("pingresult"),[self.scriptValue.encode()])
        self.assertEqual(res,'ok')

    def setupForPingRuner(self):
        self.hosts = []
        interfaces = [
        {'hostid': '1','main': '1','type': '1','ip': '127.0.0.127'},
        {'hostid': '2','main': '0','type': '1','ip': '127.0.0.1'}
        ]
        for i in range(3):
            self.hosts.append(zabmodule.initHost())
        for i,item in enumerate(self.hosts):
            interfaces[0].update([("hostid",str(i))])
            interfaces[0].update([("interfaceid",str(i)+'1')])
            interfaces[1].update([("hostid",str(i))])
            interfaces[1].update([("interfaceid",str(i)+'2')])
            item.update([('interfaces',interfaces),('hostid',str(i)),('host','id-'+str(i))])



    def testFromPingRuner(self):
        self.setupForPingRuner()
        res = zabmodule.PingRuner(None,self.hosts)
        self.assertEqual(res,'api is None')
        res = zabmodule.PingRuner(self.api,None)
        self.assertEqual(res,'hosts is None')
        res = zabmodule.PingRuner(self.api,[])
        self.assertEqual(res,'hosts is empty')
        res = zabmodule.PingRuner(self.api,self.hosts)
        for item in self.hosts:
            res = item['pingresult']
            self.assertEqual(res[0],self.scriptValue.encode())
            l = len(res[1].splitlines())
            self.assertEqual(l,8)


zabcombain/zabmodule.py
import time
from threading import Thread
from subprocess import Popen,PIPE
from os import name as osname
import queue



def initHost():
    return {'hostid': '',
            'host': '',
            'name': '',
            'description': '',
            'pingresult':'',
            'groups': [{'groupid': '', 'name': '', 'internal': '', 'flags': ''}],
            'interfaces': [{'interfaceid': '',
                            'hostid': '',
                            'main': '',
                            'type': '',
                            'useip': '',
                            'ip': '',
                            'dns': '',
                            'port': '',
                            'bulk': ''}]}

def initHostsFromData(data):
    hosts = []
    for line in data:
        i=0
        if hosts == []:
            host = initHost()
            host.update([("name",line[0]),("issues",[{"issue":line[1],"agetime":line[2]}])])
            hosts.append(host)
        else:
            for item in hosts:
                if item.get("name")==line[0]:
                    item.get("issues").append({"issue":line[1],"agetime":line[2]})
                    i+=1
            if i==0:
                host = initHost()
                host.update([("name",line[0]),("issues",[{"issue":line[1],"agetime":line[2]}])])
                hosts.append(host)
                i=0
    return hosts

def getApi(hostname,login=None,password=None):
        if not login:
             return "non login",None
        if not password:
             return "non password",None
        from pyzabbix import ZabbixAPI, ZabbixAPIException
        zapi = None
        try:
            zapi = ZabbixAPI(server = hostname)
        except Exception as e:
            return str(e),None
        try:
            zapi.login(login,password)
        except Exception as e:
            return str(e),None
        return "server connect",zapi

def initHostsFromServer(api,hosts):
    output=["hostid","host","name","description","groups","interfaces"]
    hostnames = []
    for host in hosts:
        hostnames.append(host.get("name"))
    for item in api.host.get(output=output,filter={"name":hostnames},selectInterfaces="extend",selectGroups="extend"):
        for host in hosts:
            if host.get("name")==item.get("name"):
                host.update(item.items())

def getItems(text):
    if text:
        for line in text.split("\n"):
            items = line.split("\t")
            if len(items)!=1:
                yield items


def runRemoteServerScript(api,host):
    if api is None:
        print('api is None')
    if host is None:
        print('host is None')
    result = None
    try:
        result = api.script.execute(hostid=host.get("hostid"),scriptid="1")
    except Exception as err:
        print(err)
    host.update([("pingresult",result.get("value"))])

def pingFromOS(ip):
    if ip is None:
        return b"ip is None"
    if ip=='':
        return b'ip is empty'
    if osname == "nt":
        pcArg=['ping','-n','3',ip]
    else:
        pcArg=['ping','-c','3',ip]
    proc = Popen(pcArg,stdout=PIPE)
    out = proc.stdout.readlines()
    res = b""
    for row in out:
        res+=row
    proc.stdout.close()
    return res

class TaskRuner():
    def __init__(self):
        self.tasks = queue.Queue()
        self.threads = []
        self.items = None

    def do_start(self):
        pass

    def do_work(self,*args, **kwargs):
        pass

    def do_end(self):
        pass

    def worker(self):
        while True:
            item = self.tasks.get()
            if item is None:
                break
            self.do_work(item)
            self.tasks.task_done()

    def run(self):
        self.do_start()
        for i in range(len(self.items)):
            t = Thread(target=self.worker)
            t.start()
            self.threads.append(t)

        for item in self.items:
            self.tasks.put(item)

        # block until all tasks are done
        self.tasks.join()

        # stop workers
        for i in range(len(self.items)):
            self.tasks.put(None)
        for t in self.threads:
            t.join()
        self.do_end()

def pingRuner(api,hosts):

    def iface_work(item):
        res = pingFromOS(item['ip'])
        if osname == "nt":
            res = res.decode("cp866")
        else:
            res = res.decode()
        item.update([('res',res)])

    def host_work(item):
        runRemoteServerScript(api,item)
        ifacesTasks = TaskRuner()
        ifacesTasks.items = item['interfaces']
        ifacesTasks.do_work = iface_work
        ifacesTasks.run()


    runHosts = TaskRuner()
    runHosts.items = hosts
    runHosts.do_work = host_work
    runHosts.run()


zabcombain/zabcombain.py
import sys
from PyQt5.QtWidgets import *
from PyQt5.QtGui import QFont,QIcon
from PyQt5.QtCore import *
from .zabmodule import *
from threading import Thread,Lock
from os import getcwd,path,environ,name as osname
import configparser
from datetime import datetime


def getlogin():
    name = ''
    if osname == "nt":
        name = environ['username']
    else:
        name = environ['LOGNAME']
    return name


class Page(QWidget):
    def __init__(self,*args,**kwargs):
        super().__init__(*args,**kwargs)
        # print(self.parentWidget())
        self.parent = self.parentWidget().parentWidget()
        self.settings = self.parentWidget().parentWidget().settings
        self.console  = self.parentWidget().parentWidget().console


    def consoleLog(self,text):
        dt = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        self.console.appendPlainText(dt + ' ' + text)

class PageMain(Page):
    """
    docstring PageMain
    """
    def __init__(self,*args,**kwargs):
        super().__init__(*args,**kwargs)
        self.loginfield = QLineEdit(self)
        self.loginfield.setText(getlogin())
        self.loginfield.setGeometry(10,10,100,25)
        self.passfield = QLineEdit(self)
        self.passfield.setGeometry(10,40,100,25)
        self.passfield.setEchoMode(2)
        self.loginButton = QPushButton('login',self)
        self.loginButton.setGeometry(10,70,100,25)
        self.loginButton.pressed.connect(self.OnPressLogin)
        self.settingButton = QPushButton('settings',self)
        self.settingButton.setGeometry(10,100,100,25)
        self.settingButton.pressed.connect(self.OnPressSetting)

    def OnPressLogin(self):
        error,self.parent.api = getApi(self.settings.hostname, self.loginfield.text(),self.passfield.text())
        self.consoleLog(error)

    def OnPressSetting(self):
        text, ok = QInputDialog.getMultiLineText(self, 'Settings','Write otions',self.settings.readValue)
        if ok:
            self.settings.saveSetting(str(text))
            self.settings.readSetting()

class WorkerOrder(QThread):
        finished = pyqtSignal(str)
        def __init__(self,parent = None ):
            super(WorkerOrder,self).__init__(parent)

        def setAPI(self,api):
            self.api = api

        def setHosts(self,hosts):
            self.hosts = hosts

        def getDataPing(self):
            initHostsFromServer(self.api,self.hosts)
            pingRuner(self.api,self.hosts)
            res = ''
            for host in self.hosts:
                res+='{:<9}:{}\n'.format('name',host.get('name'))
                res+='{:<9}:{}\n'.format('hostname',host.get('host'))
                if host.setdefault('issues',False):
                    for issue in host.get('issues'):
                        res+='{:<9}:{}\n'.format('issue',issue.get('issue'))
                        res+='{:<9}:{}\n'.format('agetime',issue.get('agetime'))
                res+='\nZABBIX PING RESULT\n\n'
                res+=host['pingresult']
                res+='\nINTERFACES PING RESULT\n\n'
                for item in host['interfaces']:
                    res += item['res']
                    res+='\n'
            return res

        def run(self):
            self.finished.emit(self.getDataPing())

class PageOrder(Page):
    def __init__(self,*args,**kwargs):
        super().__init__(*args,**kwargs)
        w = self.parentWidget().geometry().width()
        h = self.parentWidget().geometry().height()
        # Zabbix api - self.api
        self.api = None
        self.inputLabel = QLabel('Input',self)
        self.inputLabel.move(5,5)
        self.inputField = QPlainTextEdit(self)
        self.inputField.setGeometry(5,25,w-80,160)
        self.pasteButton = QPushButton('Paste',self)
        self.pasteButton.setGeometry(w-70,25,60,25)
        self.proccButton = QPushButton('Proccess',self)
        self.proccButton.setGeometry(w-70,55,60,25)
        self.inputCheckBox =  QCheckBox('Contin.',self)
        self.inputCheckBox.move(w-70,85)
        self.pasteButton.pressed.connect(self.OnPressPaste)
        self.proccButton.pressed.connect(self.OnPressProcess)
        # Output
        self.outputLabel = QLabel('Output',self)
        self.outputLabel.move(5,210)
        self.outputField = QPlainTextEdit(self)
        self.outputField.setGeometry(5,230,w-80,160)
        self.copyButton = QPushButton('Copy',self)
        self.copyButton.setGeometry(w-70,230,60,25)
        self.copyButton.pressed.connect(self.OnSelectCopy)


    def OnPressPaste(self):
        if self.inputField.canPaste():
            if not self.inputCheckBox.checkState():
                self.inputField.clear()
            self.inputField.paste()
            self.inputField.appendPlainText('\n')
        else:
            self.consoleLog('Buffer is empty')

    def OnPressProcess(self):
        if not self.parent.api:
            self.consoleLog('no connection to server')
        else:
            try:
                if self.wo.isFinished():
                    self.process()
                else:
                    self.consoleLog('please wait')
            except AttributeError:
                self.process()

    def process(self):
        self.hosts = initHostsFromData(self.getItems())
        self.consoleLog('Order process')
        self.consoleLog('Hosts = %d'%len(self.hosts))
        for i,v in enumerate(self.hosts):
            self.consoleLog('%d %s'%(i+1,v.get('name')))
        try:
            self.wo = WorkerOrder()
            self.wo.setAPI(self.parent.api)
            self.wo.setHosts(self.hosts)
            self.wo.start()
            self.wo.finished.connect(self.getWorkerResult)
        except Exception as er:
            print('error start Thread',er)

    def getWorkerResult(self,res):
        self.outputField.clear()
        self.outputField.appendPlainText(res)

    def OnSelectCopy(self):
        if self.outputField.toPlainText()=='':
            self.consoleLog('Output is empty')
        else:
            self.outputField.selectAll()
            self.outputField.copy()
            self.consoleLog('all text copied')

    def getItems(self):
        text = str(self.inputField.toPlainText())
        if text:
            for line in text.split('\n'):
                items = line.split('\t')
                if len(items)!=1:
                    yield items

class PagePing(Page):
    def __init__(self,*args,**kwargs):
        super().__init__(*args,**kwargs)
        w = self.parentWidget().geometry().width()
        h = self.parentWidget().geometry().height()
        self.groupLb = QLabel('Groups',self)
        self.groupLb.move(5,5)
        self.selectGroup = QComboBox(self)
        self.selectGroup.setGeometry(5,25,220,25)
        self.selectGroup.setEnabled(False)
        self.selectGroup.currentIndexChanged.connect(self.OnSelectGroup)
        self.hostLb = QLabel('Hosts',self)
        self.hostLb.move(5,55)
        self.selectHost = QComboBox(self)
        self.selectHost.setGeometry(5,75,220,25)
        self.selectHost.setEnabled(False)
        self.selectGroup.currentIndexChanged.connect(self.OnSelectHost)
        # update button
        self.updateBtn = QPushButton('Update',self)
        self.updateBtn.setGeometry(230,25,110,25)
        self.updateBtn.pressed.connect(self.onPressUpdate)
        # ping button
        self.pingButton = QPushButton('Ping',self)
        self.pingButton.setGeometry(5,110,100,25)
        self.pingButton.setEnabled(False)
        self.pingButton.pressed.connect(self.OnPressPing)
        # output field
        self.outputLb = QLabel('Output',self)
        self.outputLb.move(5,210)
        self.outputField = QPlainTextEdit(self)
        self.outputField.setGeometry(5,230,w-80,160)

    def onPressUpdate(self):
        if not self.parent.api:
            self.consoleLog('no connection to server')
        else:
            self.groups = dict(self.getGroups())
            items = list(self.groups.keys())
            items.sort()
            self.selectGroup.addItems(items)
            self.selectGroup.setEnabled(True)

    def getGroups(self):
        for item in self.parent.api.hostgroup.get():
            yield item['name'],item['groupid']

    def getHosts(self, groupid = None):
        output=['hostid','host','name','description','groups','interfaces']
        for item in self.parent.api.host.get(groupids=[groupid],filter={'status':'0'},output=output,selectInterfaces='extend'):
            yield item['name'],item

    def OnSelectGroup(self,index):
        key = self.selectGroup.currentText()
        self.hosts = dict(self.getHosts(self.groups[key]))
        items = list(self.hosts.keys())
        items.sort()
        self.selectHost.clear()
        self.selectHost.addItems(items)
        self.selectHost.setEnabled(True)

    def OnSelectHost(self,index):
        self.pingButton.setEnabled(True)

    def OnPressPing(self):
        if not self.parent.api:
            self.consoleLog('no connection to server')
        else:
            try:
                if self.wo.isFinished():
                    self.process()
                else:
                    self.consoleLog('please wait')
            except AttributeError:
                self.process()

    def process(self):
        key = self.selectHost.currentText()
        host = initHost()
        host.update(self.hosts[key].items())
        self.consoleLog('Ping for '+host.get('name'))
        try:
            self.wo = WorkerOrder()
            self.wo.setAPI(self.parent.api)
            self.wo.setHosts([host])
            self.wo.start()
            self.wo.finished.connect(self.getWorkerResult)
        except Exception as er:
            print('error start Thread',er)

    def getWorkerResult(self,res):
        self.outputField.clear()
        self.outputField.appendPlainText(res)

class Settings():
    def __init__(self):
        self.defaultValue = '[server]\nhost = http://127.0.0.1/zabbix\n'
        if osname == 'nt':
            self.home = path.join('c:\\',environ['HOMEPATH'])
        else:
            self.home = environ['HOME']
        self.here = path.join(self.home,'.zabcombain')
        self.hostname = None
        self.readSetting()

    def readSetting(self):
        if path.exists(self.here):
            with open(self.here,'r') as f:
                self.readValue = f.read()
        else:
            self.readValue=self.defaultValue
            with open(self.here,'w') as f:
                f.write(self.defaultValue)
        self.parseSetting()

    def saveSetting(self,value):
        if self.readValue != value:
            with open(self.here,'w') as f:
                f.write(value)
            self.readValue = value

    def parseSetting(self):
        config = configparser.ConfigParser()
        config.read(self.here)
        self.hostname = config['server']['host']

class MainWindow(QMainWindow):
    def __init__(self,*args,**kwargs):
        super(MainWindow,self).__init__(*args,**kwargs)
        self.api = None
        self.settings = Settings()
        self.initUI()

    def initUI(self):
        self.width = 600
        self.height = 600
        self.resize(self.width,self.height)
        QToolTip.setFont(QFont('SansSerif',10))
        self.setToolTip('Zabcombain')
        self.console = QPlainTextEdit(self)
        self.console.setGeometry(5,450,self.width-10,120)
        self.panel = QTabWidget(self)
        self.panel.setGeometry(5,5,self.width-10,440)
        self.panel.addTab(PageMain(self.panel),'Main')
        self.panel.addTab(PageOrder(self.panel),'Oreders')
        self.panel.addTab(PagePing(self.panel),'Ping')
        self.center()
        here = path.abspath(path.dirname(__file__))
        here = path.split(here)[0]
        icon_file = path.join(here,'data','favicon.ico')
        icon = QIcon(path.abspath(icon_file))
        self.setWindowIcon(icon)
        self.setWindowTitle('Zabcombain')
        self.show()

    def center(self):
        qr = self.frameGeometry()
        cp = QDesktopWidget().availableGeometry().center()
        qr.moveCenter(cp)
        self.move(qr.topLeft())

if __name__ == '__main__':
    # Next, create an application object.
    QApplication.setDesktopSettingsAware(False)
    app = QApplication(sys.argv)
    # Show it.
    frame = MainWindow(None)
    # Start the event loop.
    sys.exit(app.exec_())
